presto-main/src/main/java/com/facebook/presto/operator/aggregation/noisyaggregation/sketch/Bitmap.java

```diff
@@ -0,0 +1,172 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.noisyaggregation.sketch;
+
+import io.airlift.slice.SliceInput;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.BitSet;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A level of abstraction over the bitmaps used in sketches such as SFM.
+ * Abstractly, these are essentially fixed-length arrays of booleans that support flipping and applying randomized response.
+ * This is implemented as a wrapper around Java's BitSet.
+ * <p>
+ * Note: The byte arrays in toBytes() and fromSliceInput() are variable-length.
+ * Trailing zeros are implicitly truncated in these functions.
+ * The fixed-length nature of the bitmap comes into play in flipAll (randomized response),
+ * where every bit from 0 to length-1 must be flipped with a fixed probability.
+ */
+public class Bitmap
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(Bitmap.class).instanceSize();
+    private static final int BITSET_INSTANCE_SIZE = ClassLayout.parseClass(BitSet.class).instanceSize();
+
+    private final BitSet bitSet;
+    private final int length;
+
+    public Bitmap(int length)
+    {
+        checkArgument(length > 0, "length must be positive");
+        bitSet = new BitSet(length);
+        this.length = length;
+    }
+
+    private Bitmap(int length, BitSet bitSet)
+    {
+        requireNonNull(bitSet, "bitSet cannot be null");
+        checkArgument(length >= bitSet.length(), "bitmap size must be large enough to cover existing bits");
+        this.bitSet = bitSet;
+        this.length = length;
+    }
+
+    public static Bitmap fromBytes(int length, byte[] bytes)
+    {
+        return new Bitmap(length, BitSet.valueOf(bytes));
+    }
+
+    public static Bitmap fromSliceInput(SliceInput input, int byteCount, int length)
+    {
+        checkArgument(byteCount >= 0, "byteCount must be nonnegative");
+        if (byteCount == 0) {
+            return new Bitmap(length);
+        }
+
+        byte[] bytes = new byte[byteCount];
+        input.readBytes(bytes);
+        return Bitmap.fromBytes(length, bytes);
+    }
+
+    public byte[] toBytes()
+    {
+        return bitSet.toByteArray();
+    }
+
+    /**
+     * Length of toBytes()
+     */
+    public int byteLength()
+    {
+        // per https://docs.oracle.com/javase/8/docs/api/java/util/BitSet.html#toByteArray--
+        return (bitSet.length() + 7) / 8;
+    }
+
+    @Override
+    public Bitmap clone()
+    {
+        return Bitmap.fromBytes(length, bitSet.toByteArray());
+    }
+
+    public long getRetainedSizeInBytes()
+    {
+        // Under the hood, BitSet stores a long[] array of BitSet.size() bits
+        return INSTANCE_SIZE + BITSET_INSTANCE_SIZE + bitSet.size() / Byte.SIZE;
```

Comment:
nit: `sizeOfLongArray(bitSet.size() / Long.SIZE)` to account for the `long[]` object headers.